File: frontend/src/api/categories.ts
import { authorizedRequest } from './authorizedRequest'
import type { Category } from '../types/task'

export function fetchCategories(): Promise<Category[]> {
  return authorizedRequest<Category[]>('/v1/categories')
}

export interface CategoryInput {
  name: string
  color: string
}

export function createCategory(input: CategoryInput): Promise<Category> {
  return authorizedRequest<Category>('/v1/categories', { method: 'POST', body: input })
}

export function updateCategory(id: string, input: Partial<CategoryInput>): Promise<Category> {
  return authorizedRequest<Category>(`/v1/categories/${id}`, { method: 'PATCH', body: input })
}

export function deleteCategory(id: string): Promise<void> {
  return authorizedRequest<void>(`/v1/categories/${id}`, { method: 'DELETE' })
}

File: backend/app/api/v1/categories.py
import uuid

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.category import Category
from app.models.user import User
from app.schemas.category import CategoryCreate, CategoryRead, CategoryUpdate
from app.services import category_service

router = APIRouter(prefix="/categories", tags=["categories"])


@router.post("", response_model=CategoryRead, status_code=status.HTTP_201_CREATED)
async def create_category(
    category_in: CategoryCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Category:
    return await category_service.create_category(db, current_user.id, category_in)


@router.get("", response_model=list[CategoryRead])
async def list_categories(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[Category]:
    return await category_service.list_categories(db, current_user.id)


@router.get("/{category_id}", response_model=CategoryRead)
async def get_category(
    category_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Category:
    return await category_service.get_category(db, current_user.id, category_id)


@router.patch("/{category_id}", response_model=CategoryRead)
async def update_category(
    category_id: uuid.UUID,
    category_in: CategoryUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> Category:
    return await category_service.update_category(db, current_user.id, category_id, category_in)


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_category(
    category_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> None:
    await category_service.delete_category(db, current_user.id, category_id)
